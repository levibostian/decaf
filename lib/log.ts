import colors from 'npm:ansi-styles@6.2.1'

// log.ts
// This module provides a simple API for logging messages at different levels to GitHub Actions.

// GitHub Actions supports setting log level by prefixing the message with specific tokens.
// Reference: https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-a-debug-message

// Log levels
const levels = {
  debug: `::debug::${colors.white.open}`, // Debug messages, not shown by default in GitHub Actions logs.
  warning: `::warning::${colors.white.open}`, // Lines are highlighted with yellow in the GitHub Actions logs.
  error: `::error::${colors.white.open}`, // Lines are highlighted with red in the GitHub Actions logs.
  notice: `${colors.blue.open}`, // Notice messages, a way to highlight important information, displayed in blue.
  message: `${colors.white.open}`, 
};

// Generic log function
function log(level: keyof typeof levels, message: string) {
  console.log(`${levels[level]}${message}`);
}

/**
 * Logs a debug message, which is not shown by default but can be enabled in the GitHub Actions workflow settings.
 * Debug messages are useful for detailed troubleshooting information.
 */
export function debug(message: string) {
  log("debug", message);
}

/**
 * Logs a warning message, displayed in yellow in the GitHub Actions logs.
 * Warning messages are useful for non-critical issues that should be highlighted to users.
 */
export function warning(message: string) {
  log("warning", message);
}

/**
 * Logs an error message, displayed in red in the GitHub Actions logs.
 * Error messages are critical and indicate something went wrong during the execution of the workflow.
 */
export function error(message: string) {
  log("error", message);
}

/**
 * Logs a notice message, displayed in blue in the GitHub Actions logs.
 * Notice messages are useful for highlighting important information that is not necessarily an error or warning.
 */
export function notice(message: string) {
  log("notice", message);
}

/**
 * Logs a standard message without any specific log level prefix.
 * These messages are displayed in the standard log color and are useful for general information.
 */
export function message(message: string) {
  log("message", message);
}
