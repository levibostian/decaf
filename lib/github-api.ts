import * as log from "./log.ts"

export interface GitHubRelease {
  tag: {
    name: string
    commit: {
      sha: string
    }
  }
  name: string
  created_at: Date
}

export interface GitHubCommit {
  sha: string
  message: string
  date: Date
}

interface GitHubCommitApiResponse {
  sha: string
  commit: {
    message: string
    committer: {
      date: string
    }
  }
}

export interface GitHubPullRequest {
  prNumber: number
  targetBranchName: string
  sourceBranchName: string
  title: string
  description: string
}

// Returns list of all open pull requests that are stacked on top of each other.
// Index 0 is the newest pull request.
const getPullRequestStack = async (
  { owner, repo, startingBranch }: { owner: string; repo: string; startingBranch: string },
): Promise<GitHubPullRequest[]> => {
  // get list of all open pull requests.
  const graphqlQuery = `
query($owner: String!, $repo: String!, $endCursor: String, $numberOfResults: Int!) {
  repository(owner: $owner, name: $repo) {
    pullRequests(first: $numberOfResults, states: [OPEN], after: $endCursor, orderBy: {field: CREATED_AT, direction: DESC}) {
      nodes {
        prNumber: number
        targetBranchName: baseRefName
        sourceBranchName: headRefName     
        title: title
        description: body # description in markdown. you can do html or text instead. 
      }
      pageInfo {        
        endCursor
        hasNextPage
      }
    }
  }
}
`

  let pullRequests: GitHubPullRequest[] = []

  await githubGraphqlRequestPaging<{
    data: {
      repository: {
        pullRequests: {
          nodes: {
            prNumber: number
            targetBranchName: string
            sourceBranchName: string
            title: string
            description: string
          }[]
          pageInfo: {
            endCursor: string
            hasNextPage: boolean
          }
        }
      }
    }
  }>(
    graphqlQuery,
    { owner, repo, numberOfResults: 100 },
    async (response) => {
      pullRequests.push(...response.data.repository.pullRequests.nodes)
      return true
    },
  )

  // Takes the list of all pull requests for the repo and puts into a list, starting at the startingBranch PR, and creates the stack. Going from starting to the top of the stack (example: PR 1 -> PR 2 -> PR 3).
  const startingPullRequest = pullRequests.find((pr) => pr.sourceBranchName === startingBranch)
  if (!startingPullRequest) {
    throw new Error(
      `Could not get pull request stack because not able to find pull request for starting branch, ${startingBranch}. This is unexpected.`,
    )
  }

  const prStack: GitHubPullRequest[] = [startingPullRequest]
  let sourceBranchSearchingFor = startingBranch

  while (true) {
    const nextPullRequest = pullRequests.find((pr) => pr.targetBranchName === sourceBranchSearchingFor)

    if (!nextPullRequest) {
      break
    }

    prStack.push(nextPullRequest)
    sourceBranchSearchingFor = nextPullRequest.sourceBranchName
  }

  return prStack
}

// Get a list of github releases.
// The github rest api does not return the git tag (and commit sha) for a release. You would need to also call the tags endpoint and match the tag name to the release name (painful).
// But I found you can use the github graphql api to get this information in 1 call.
const getTagsWithGitHubReleases = async (
  { owner, repo, processReleases, numberOfResults }: {
    owner: string
    repo: string
    processReleases: (data: GitHubRelease[]) => Promise<boolean>
    numberOfResults?: number
  },
): Promise<void> => {
  // Gets list of tags that also have a github release made for it.
  // If a tag does not have a release, it will not be returned.
  // Sorted by latest release first.
  // Paging enabled.
  const graphqlQuery = `
query($owner: String!, $repo: String!, $endCursor: String, $numberOfResults: Int!) {
  repository(owner: $owner, name: $repo) {
    releases(first: $numberOfResults, after: $endCursor) {
      nodes {
        name # name of github release 
        createdAt # "2024-06-06T04:26:30Z"
        isDraft # true if release is a draft
        tag {
          name # tag name 
          target {
            ... on Commit {
              oid # commit sha hash
            }
          }
        }
      }
      pageInfo {
        endCursor
        hasNextPage
      }
    }
  }
}
`

  await githubGraphqlRequestPaging<{
    data: {
      repository: {
        releases: {
          nodes: {
            name: string
            createdAt: string
            isDraft: boolean
            tag: {
              name: string
              target: {
                oid: string
              }
            }
          }[]
          pageInfo: {
            endCursor: string
            hasNextPage: boolean
          }
        }
      }
    }
  }>(
    graphqlQuery,
    { owner, repo, numberOfResults: numberOfResults || 100 },
    (response) => {
      const releases: GitHubRelease[] = response.data.repository.releases.nodes
        .filter((release) => !release.isDraft) // only look at releases that are not drafts
        .map((release) => {
          return {
            tag: {
              name: release.tag.name,
              commit: {
                sha: release.tag.target.oid,
              },
            },
            name: release.name,
            created_at: new Date(release.createdAt),
          }
        })

      return processReleases(releases)
    },
  )
}

const getCommitsForBranch = async <T>(
  { owner, repo, branch, processCommits }: {
    owner: string
    repo: string
    branch: string
    processCommits: (data: GitHubCommit[]) => Promise<boolean>
  },
) => {
  return await githubApiRequestPaging<GitHubCommitApiResponse[]>(
    `https://api.github.com/repos/${owner}/${repo}/commits?sha=${branch}&per_page=100`,
    async (apiResponse) => {
      return await processCommits(apiResponse.map((response) => {
        return {
          sha: response.sha,
          message: response.commit.message,
          date: new Date(response.commit.committer.date),
        }
      }))
    },
  )
}

const createGitHubRelease = async (
  { owner, repo, tagName, commit }: {
    owner: string
    repo: string
    tagName: string
    commit: GitHubCommit
  },
) => {
  await githubApiRequest(
    `https://api.github.com/repos/${owner}/${repo}/releases`,
    "POST",
    {
      tag_name: tagName,
      target_commitish: commit.sha,
      name: tagName,
      body: "",
      draft: false,
      prerelease: false,
    },
  )
}

// Make a GitHub Rest API request.
const githubApiRequest = async <T>(
  url: string,
  method: "GET" | "POST" = "GET",
  body: object | undefined = undefined,
) => {
  const headers = {
    "Authorization": `Bearer ${Deno.env.get("INPUT_GITHUB_TOKEN")}`,
    "Accept": "application/vnd.github.v3+json",
    "Content-Type": "application/json",
  }

  log.debug(
    `GitHub API request: ${method}:${url}, headers: ${JSON.stringify(headers)}, body: ${JSON.stringify(body)}`,
  )

  const response = await fetch(url, {
    method,
    headers,
    body: body ? JSON.stringify(body) : undefined,
  })

  if (!response.ok) {
    throw new Error(
      `Failed to call github API endpoint: ${url}, given error: ${response.statusText}`,
    )
  }

  const responseJsonBody: T = await response.json()

  return {
    status: response.status,
    body: responseJsonBody,
    headers: response.headers,
  }
}

// Make a GitHub Rest API request that supports paging. Takes in a function that gets called for each page of results.
// In that function, return true if you want to get the next page of results.
async function githubApiRequestPaging<RESPONSE>(
  initialUrl: string,
  processResponse: (data: RESPONSE) => Promise<boolean>,
): Promise<void> {
  let url = initialUrl
  let getNextPage = true

  while (getNextPage) {
    const response = await githubApiRequest<RESPONSE>(url)

    getNextPage = await processResponse(response.body)

    // for propagation, add nextLink to responseJsonBody. It's the URL that should be used in the next HTTP request to get the next page of results.
    const linkHeader = response.headers.get("Link")?.match(
      /<(.*?)>; rel="next"/,
    )
    const nextPageUrl = linkHeader ? linkHeader[1] : undefined

    if (!nextPageUrl) {
      getNextPage = false
    } else {
      url = nextPageUrl
    }
  }
}

/*

  // Make a GitHub GraphQL API request.

  Example:
  // const QUERY = `
  // query($owner: String!, $name: String!) {
  //   repository(owner: $owner, name: $name) {
  //     releases(first: 100, orderBy: {field: CREATED_AT, direction: DESC}) {
  //       nodes {
  //         name
  //         createdAt
  //         tag {
  //           name
  //           target {
  //             ... on Commit {
  //               oid
  //             }
  //           }
  //         }
  //       }
  //     }
  //   }
  // }
  // `;
  // const variables = {
  //   owner: 'REPO_OWNER', // Replace with the repository owner
  //   name: 'REPO_NAME'    // Replace with the repository name
  // };
*/
const githubGraphqlRequest = async <T>(query: string, variables: object) => {
  const headers = {
    "Authorization": `Bearer ${Deno.env.get("INPUT_GITHUB_TOKEN")}`,
    "Content-Type": "application/json",
  }

  const body = JSON.stringify({
    query,
    variables,
  })

  log.debug(
    `GitHub graphql request: headers: ${JSON.stringify(headers)}, body: ${body}`,
  )

  const response = await fetch("https://api.github.com/graphql", {
    method: "POST",
    headers,
    body,
  })

  if (!response.ok) {
    throw new Error(
      `Failed to call github graphql api. Given error: ${response.statusText}`,
    )
  }

  const responseJsonBody: T = await response.json()

  log.debug(`GitHub graphql response: ${JSON.stringify(responseJsonBody)}`)

  return {
    status: response.status,
    body: responseJsonBody,
    headers: response.headers,
  }
}

// Make a GitHub GraphQL API request that supports paging. Takes in a function that gets called for each page of results.
// Assumptions when using this function:
// 1. Query must contain a pageInfo object:
// pageInfo {
//  endCursor
//  hasNextPage
// }
// 2. Query contains a variable called endCursor that is used to page through results.
// See: https://docs.github.com/en/graphql/guides/using-pagination-in-the-graphql-api to learn more about these assumptions.
async function githubGraphqlRequestPaging<RESPONSE>(
  query: string,
  variables: { [key: string]: string | number },
  processResponse: (data: RESPONSE) => Promise<boolean>,
): Promise<void> {
  // deno-lint-ignore no-explicit-any
  function findPageInfo(
    _obj: any,
  ): { hasNextPage: boolean; endCursor: string } {
    // Create a shallow copy of the object to avoid modifying the original
    const obj = { ..._obj }

    // nodes is the JSON response. It could be a really big object. Do not perform recursion on it, so let's delete it.
    delete obj["nodes"]

    for (const key in obj) {
      if (key === "pageInfo") {
        return obj[key] as { hasNextPage: boolean; endCursor: string }
      } else {
        return findPageInfo(obj[key])
      }
    }

    throw new Error(
      "pageInfo object not found in response. Did you forget to add pageInfo to your graphql query?",
    )
  }

  let getNextPage = true
  while (getNextPage) {
    const response = await githubGraphqlRequest<RESPONSE>(query, variables)

    getNextPage = await processResponse(response.body)

    const pageInfo = findPageInfo(response.body)

    log.debug(`pageInfo: ${JSON.stringify(pageInfo)}`)

    if (!pageInfo.hasNextPage) {
      getNextPage = false
    } else {
      variables["endCursor"] = pageInfo.endCursor
    }
  }
}

export interface GitHubApi {
  getTagsWithGitHubReleases: typeof getTagsWithGitHubReleases
  getCommitsForBranch: typeof getCommitsForBranch
  createGitHubRelease: typeof createGitHubRelease
  getPullRequestStack: typeof getPullRequestStack
}

export const GitHubApiImpl: GitHubApi = {
  getTagsWithGitHubReleases,
  getCommitsForBranch,
  createGitHubRelease,
  getPullRequestStack,
}
