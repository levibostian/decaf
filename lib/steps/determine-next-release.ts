import { GitHubCommit, GitHubRelease } from "../github-api.ts"
import * as semver from "@std/semver"
import { versionBumpForCommitBasedOnConventionalCommit } from "../conventional-commits.ts"
import { Logger, logger } from "../log.ts"
import { GetNextReleaseVersionEnvironment } from "../types/environment.ts"

export interface DetermineNextReleaseStepConfig {
  branches?: {
    branch_name?: string
    prerelease?: boolean
    version_suffix?: string
  }[]
}

export interface DetermineNextReleaseStep {
  getNextReleaseVersion({ config, environment, commits, latestRelease }: {
    config?: DetermineNextReleaseStepConfig
    environment: GetNextReleaseVersionEnvironment
    commits: GitHubCommit[]
    latestRelease: GitHubRelease | null
  }): Promise<string | null>
}

export class DetermineNextReleaseStepImpl implements DetermineNextReleaseStep {
  private log: Logger

  constructor(log: Logger = logger) {
    this.log = log
  }

  async getNextReleaseVersion({ config, environment, commits, latestRelease }: {
    config?: DetermineNextReleaseStepConfig
    environment: GetNextReleaseVersionEnvironment
    commits: GitHubCommit[]
    latestRelease: GitHubRelease | null
  }): Promise<string | null> {
    // First, parse all commits to determine the version bump for each commit.
    const versionBumpsForEachCommit = commits.map((commit) => {
      const firstLineOfCommitMessage = commit.message.split("\n")[0]
      const abbreviatedFirstLineOfCommitMessage = firstLineOfCommitMessage.length > 50
        ? firstLineOfCommitMessage.substring(0, 50) + "..."
        : firstLineOfCommitMessage
      const first8CharactersOfCommitHash = commit.sha.substring(0, 8)

      const versionBumpForCommit = versionBumpForCommitBasedOnConventionalCommit(
        commit,
      )

      const logPrefix = `${abbreviatedFirstLineOfCommitMessage} (${first8CharactersOfCommitHash})`
      switch (versionBumpForCommit) {
        case "major":
          this.log.message(`${logPrefix} => indicates a major release.`)
          break
        case "minor":
          this.log.message(`${logPrefix} => indicates a minor release.`)
          break
        case "patch":
          this.log.message(`${logPrefix} => indicates a patch release.`)
          break
        default:
          this.log.message(`${logPrefix} => does not indicate a release.`)
          break
      }

      return versionBumpForCommit
    }).filter((versionBump) => versionBump !== undefined) as ("patch" | "major" | "minor")[]

    // If none of the commits indicate a release should be made, exit early.
    if (versionBumpsForEachCommit.length === 0) {
      return null
    }

    const lastReleaseVersion = latestRelease?.tag.name
    const isNextReleasePrerelease = config?.branches?.find((branch) => branch.branch_name === environment.gitCurrentBranch)?.prerelease
    const prereleaseVersionSuffix = config?.branches?.find((branch) => branch.branch_name === environment.gitCurrentBranch)?.version_suffix ||
      environment.gitCurrentBranch

    // If there was not a last release version, then this is the first release. Return a version to start with.
    if (!lastReleaseVersion) {
      if (isNextReleasePrerelease) return `1.0.0-${prereleaseVersionSuffix}.1`
      else return "1.0.0"
    }

    const lastReleaseSemanticVersion = semver.tryParse(lastReleaseVersion)
    if (!lastReleaseSemanticVersion) {
      throw new Error(
        `The last release version, ${lastReleaseVersion}, is not a valid semantic version (https://semver.org/). I can only determine the next release version if the latest release is a valid semantic version. Push a new valid version and try again.`,
      )
    }

    // Common code to get the next semantic version for a given bump type.
    // Code is heavily inspired by semantic-release's implementation to get the next version.
    // https://github.com/semantic-release/semantic-release/blob/45bf9d601591bf7649926e54a9459c643136b485/lib/get-next-version.js
    // the unit tests for this file is the best reference to understand this code.
    const getNextSemanticVersionForBump = (bump: "major" | "minor" | "patch") => {
      if (isNextReleasePrerelease) {
        const isLatestReleaseSameSuffix = lastReleaseSemanticVersion.prerelease && lastReleaseSemanticVersion.prerelease.length > 0 &&
          lastReleaseSemanticVersion.prerelease[0] === prereleaseVersionSuffix

        // If there is the same suffix, we have 2 use cases to handle.
        // 1. Given 1.1.0-beta.2 and bump is minor or patch, we should increment the prerelease version to 1.1.0-beta.3
        // 2. Given 1.1.0-beta.2 and bump is major, we should increment the major version to 2.0.0-beta.1
        if (isLatestReleaseSameSuffix) {
          // Generate both of these use cases and return the greater version which resolves the conflict.
          const version1 = semver.increment(lastReleaseSemanticVersion, "prerelease")
          const version2 = semver.parse(`${semver.format(semver.increment(lastReleaseSemanticVersion, bump))}-${prereleaseVersionSuffix}.1`)

          return semver.format(semver.greaterThan(version1, version2) ? version1 : version2)
        } else {
          // if the suffix changes, we perform the version bump, add new suffix, and reset the prerelease version to 1.
          return `${semver.format(semver.increment(lastReleaseSemanticVersion, bump))}-${prereleaseVersionSuffix}.1`
        }
      } else {
        return semver.format(semver.increment(lastReleaseSemanticVersion, bump))
      }
    }

    // return the next release version based on the type of bump indicated by the commits. Prioritize major, then minor, then patch.
    if (versionBumpsForEachCommit.includes("major")) {
      return getNextSemanticVersionForBump("major")
    } else if (versionBumpsForEachCommit.includes("minor")) {
      return getNextSemanticVersionForBump("minor")
    } else if (versionBumpsForEachCommit.includes("patch")) {
      return getNextSemanticVersionForBump("patch")
    }

    return null
  }
}
