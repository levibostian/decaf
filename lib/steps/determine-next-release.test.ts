import { assertEquals } from "https://deno.land/std@0.224.0/assert/assert_equals.ts";
import { GitHubCommit, GitHubRelease } from "../github-api.ts";
import { DetermineNextReleaseStepImpl } from "./determine-next-release.ts";
import { GitHubCommitFake, GitHubReleaseFake } from "../github-api.test.ts";
import { before, beforeEach, describe, it } from "jsr:@std/testing@1/bdd";
import { Logger } from "../log.ts";
import { getLogMock, LogMock } from "../log.test.ts";
import { assertSnapshot } from "jsr:@std/testing@1/snapshot";

const defaultEnvironment = {
  gitCurrentBranch: "main",
  lastRelease: null,
  gitCommitsSinceLastRelease: [],
  gitRepoOwner: "owner",
  gitRepoName: "repo",
  isDryRun: false,
};

Deno.test("given this is first release, not prerelease, expect default version", async () => {
  const commits: GitHubCommit[] = [{
    sha: "",
    message: "feat: initial commit",
    date: new Date(),
  }];
  const result = await new DetermineNextReleaseStepImpl().getNextReleaseVersion(
    {
      environment: defaultEnvironment,
      commits,
      latestRelease: null,
    },
  );
  assertEquals(result, "1.0.0");
});

Deno.test("given this is first release, prerelease, expect default prerelease version", async () => {
  const commits: GitHubCommit[] = [{
    sha: "",
    message: "feat: initial commit",
    date: new Date(),
  }];
  const result = await new DetermineNextReleaseStepImpl().getNextReleaseVersion(
    {
      config: {branches: [{branch_name: "beta", prerelease: true, version_suffix: "beta"}]},
      environment: { ...defaultEnvironment, gitCurrentBranch: "beta" },
      commits,
      latestRelease: null,
    },
  );
  assertEquals(result, "1.0.0-beta.1");
})

Deno.test("given introducing a breaking change, expect bumps major version", async () => {
  const commits: GitHubCommit[] = [{
    sha: "",
    message: "feat!: add new authentication system",
    date: new Date(),
  }];
  const result = await new DetermineNextReleaseStepImpl().getNextReleaseVersion(
    {
      environment: defaultEnvironment,
      commits,
      latestRelease: {
        ...GitHubReleaseFake,
        tag: { ...GitHubReleaseFake.tag, name: "1.2.3" },
      },
    },
  );
  assertEquals(result, "2.0.0");
});

Deno.test("given a feature commit, expect bumps minor version", async () => {
  const commits: GitHubCommit[] = [{
    sha: "",
    message: "feat: add new feature",
    date: new Date(),
  }];
  const result = await new DetermineNextReleaseStepImpl().getNextReleaseVersion(
    {
      environment: defaultEnvironment,
      commits,
      latestRelease: {
        ...GitHubReleaseFake,
        tag: { ...GitHubReleaseFake.tag, name: "1.2.3" },
      },
    },
  );
  assertEquals(result, "1.3.0");
});

Deno.test("given a fix commit, expect bumps patch version", async () => {
  const commits: GitHubCommit[] = [{
    sha: "",
    message: "fix: resolve issue with login",
    date: new Date(),
  }];
  const result = await new DetermineNextReleaseStepImpl().getNextReleaseVersion(
    {
      environment: defaultEnvironment,
      commits,
      latestRelease: {
        ...GitHubReleaseFake,
        tag: { ...GitHubReleaseFake.tag, name: "1.2.3" },
      },
    },
  );
  assertEquals(result, "1.2.4");
});

Deno.test("given a chore commit, expect no next version", async () => {
  const commits: GitHubCommit[] = [{
    sha: "",
    message: "chore: update dependencies",
    date: new Date(),
  }];
  const result = await new DetermineNextReleaseStepImpl().getNextReleaseVersion(
    {
      environment: defaultEnvironment,
      commits,
      latestRelease: {
        ...GitHubReleaseFake,
        tag: { ...GitHubReleaseFake.tag, name: "1.2.3" },
      },
    },
  );
  assertEquals(result, null);
});

Deno.test("given latest release is not prerelease and next release is prerelease, expect bump and add prerelease suffix", async () => {
  const commits: GitHubCommit[] = [{
    sha: "",
    message: "feat: add new feature",
    date: new Date(),
  }];
  const result = await new DetermineNextReleaseStepImpl().getNextReleaseVersion(
    {
      config: {branches: [{branch_name: "beta", prerelease: true, version_suffix: "beta"}]},
      environment: { ...defaultEnvironment, gitCurrentBranch: "beta" },
      commits,
      latestRelease: {
        ...GitHubReleaseFake,
        tag: { ...GitHubReleaseFake.tag, name: "1.2.3" },
      },
    },
  );
  assertEquals(result, "1.3.0-beta.1");
})

Deno.test("given latest release is prerelease, next release is prerelease, next release is major bump, expect next prerelease version with new major version", async () => {
  const commits: GitHubCommit[] = [{
    sha: "",
    message: "feat!: add new feature",
    date: new Date(),
  }];
  const result = await new DetermineNextReleaseStepImpl().getNextReleaseVersion(
    {
      config: {branches: [{branch_name: "beta", prerelease: true, version_suffix: "beta"}]},
      environment: { ...defaultEnvironment, gitCurrentBranch: "beta" },
      commits,
      latestRelease: {
        ...GitHubReleaseFake,
        tag: { ...GitHubReleaseFake.tag, name: "1.2.3-beta.1" },
      },
    },
  );
  assertEquals(result, "2.0.0-beta.1");
})

Deno.test("given latest version is prerelease and next release is prerelease, expect next prerelease version", async () => {
  const commits: GitHubCommit[] = [{
    sha: "",
    message: "feat: add new feature",
    date: new Date(),
  }];
  const result = await new DetermineNextReleaseStepImpl().getNextReleaseVersion(
    {
      config: {branches: [{branch_name: "beta", prerelease: true, version_suffix: "beta"}]},
      environment: { ...defaultEnvironment, gitCurrentBranch: "beta" },
      commits,
      latestRelease: {
        ...GitHubReleaseFake,
        tag: { ...GitHubReleaseFake.tag, name: "1.3.0-beta.1" },
      },
    },
  );
  assertEquals(result, "1.3.0-beta.2");
})

Deno.test("given latest version is prerelease and next release is not prerelease, expect next non-prelease version", async () => {
  const commits: GitHubCommit[] = [{
    sha: "",
    message: "feat: add new feature",
    date: new Date(),
  }];
  const result = await new DetermineNextReleaseStepImpl().getNextReleaseVersion(
    {
      config: {branches: [{branch_name: "beta", prerelease: true, version_suffix: "beta"}]},
      environment: { ...defaultEnvironment, gitCurrentBranch: "main" },
      commits,
      latestRelease: {
        ...GitHubReleaseFake,
        tag: { ...GitHubReleaseFake.tag, name: "1.3.0-beta.1" },
      },
    },
  );
  assertEquals(result, "1.3.0");
})

Deno.test("given latest version is prerelease and next release is prerelease but different suffix, expect next prerelease version with new suffix", async () => {
  const commits: GitHubCommit[] = [{
    sha: "",
    message: "feat: add new feature",
    date: new Date(),
  }];
  const result = await new DetermineNextReleaseStepImpl().getNextReleaseVersion(
    {
      config: {branches: [{branch_name: "beta", prerelease: true, version_suffix: "beta"}]},
      environment: { ...defaultEnvironment, gitCurrentBranch: "beta" },
      commits,
      latestRelease: {
        ...GitHubReleaseFake,
        tag: { ...GitHubReleaseFake.tag, name: "1.3.0-alpha.3" },
      },
    },
  );
  assertEquals(result, "1.3.0-beta.1");
})

describe("user facing logs", () => {
  let logMock: LogMock;

  beforeEach(() => {
    logMock = getLogMock()
  })

  it("given commit that does not trigger a release, expect logs to communicate this clearly", async (t) => {
    const commits: GitHubCommit[] = [new GitHubCommitFake(
      {
        message: "chore: does not trigger a release",
      }
    )];
    await new DetermineNextReleaseStepImpl(logMock).getNextReleaseVersion({
        environment: defaultEnvironment,
        commits,
        latestRelease: null,
      },
    );
    await assertSnapshot(t, logMock.getLogs({includeDebugLogs: false}));
  });

  it("given super long commit messages, expect logs to not be super long", async (t) => {
    const commitMessage = "feat: " + "a".repeat(1000);
    const commits: GitHubCommit[] = [new GitHubCommitFake(
      {
        message: commitMessage,
      }
    )];
    await new DetermineNextReleaseStepImpl(logMock).getNextReleaseVersion({
        environment: defaultEnvironment,
        commits,
        latestRelease: null,
      },
    );
    await assertSnapshot(t, logMock.getLogs({includeDebugLogs: false}));
  });

  it("given commit message multiple lines long, expect logs to not be super long", async (t) => {
    const commitMessage = "feat: " + "a\n".repeat(1000);
    const commits: GitHubCommit[] = [new GitHubCommitFake(
      {
        message: commitMessage,
      }
    )];
    await new DetermineNextReleaseStepImpl(logMock).getNextReleaseVersion({
        environment: defaultEnvironment,
        commits,
        latestRelease: null,
      },
    );
    await assertSnapshot(t, logMock.getLogs({includeDebugLogs: false}));
  });

  it("given many commits, expect logs to read them all nicely", async (t) => {
    const commits: GitHubCommit[] = [
      new GitHubCommitFake({message: "feat: add new feature"}),
      new GitHubCommitFake({message: "fix: resolve issue"}),
      new GitHubCommitFake({message: "chore: update dependencies"}),
    ];
    await new DetermineNextReleaseStepImpl(logMock).getNextReleaseVersion({
        environment: defaultEnvironment,
        commits,
        latestRelease: null,
      },
    );
    await assertSnapshot(t, logMock.getLogs({includeDebugLogs: false}));
  })
})
