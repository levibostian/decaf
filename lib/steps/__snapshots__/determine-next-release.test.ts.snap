export const snapshot = {};

snapshot[`user facing logs > given commit that does not trigger a release, expect logs to communicate this clearly 1`] = `
[
  "chore: does not trigger a release (abc123) => does not indicate a release.",
]
`;

snapshot[`user facing logs > given super long commit messages, expect logs to not be super long 1`] = `
[
  "feat: aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa... (abc123) => indicates a minor release.",
]
`;

snapshot[`user facing logs > given commit message multiple lines long, expect logs to not be super long 1`] = `
[
  "feat: a (abc123) => indicates a minor release.",
]
`;

snapshot[`user facing logs > given many commits, expect logs to read them all nicely 1`] = `
[
  "feat: add new feature (abc123) => indicates a minor release.",
  "fix: resolve issue (abc123) => indicates a patch release.",
  "chore: update dependencies (abc123) => does not indicate a release.",
]
`;
